lti_store/models.py

```diff
@@ -34,5 +70,141 @@ class ExternalLtiConfiguration(models.Model):
+        redirect uri's, enter them here. If you use this field you must enter all valid
+        redirect uri's the tool may request."""),
+        validators=[validate_list_field],
+    )
+    lti_1p3_public_jwk = models.JSONField(
+        "LTI 1.3 Public JWK",
+        default=dict,
+        blank=True,
+        help_text=_("Platform's generated JWK keyset."),
+    )
+    lti_advantage_deep_linking_launch_url = models.URLField(
+        "LTI Advantage Deep Linking launch URL",
+        max_length=255,
+        blank=True,
+        help_text=_("""This is the LTI Advantage Deep Linking launch URL. If the LTI Tool
+        does not provide one, use the same value as lti_1p3_launch_url."""),
+    )
+
     def __str__(self):
         return f"<ExternalLtiConfiguration #{self.id}: {self.slug}>"
+
+    def clean(self):
+        validation_errors = {}
+
+        if self.version == LTIVersion.LTI_1P1:
+            for field in [
+                "lti_1p1_launch_url",
+                "lti_1p1_client_key",
+                "lti_1p1_client_secret",
+            ]:
+                # Raise ValidationError exception for any missing LTI 1.1 field.
+                if not getattr(self, field):
+                    validation_errors.update({field: _(MESSAGES["required"])})
+
+        if self.version == LTIVersion.LTI_1P3:
+            if not self.lti_1p3_private_key:
+                # Raise ValidationError if private key is missing.
+                validation_errors.update(
+                    {"lti_1p3_private_key": _(MESSAGES["required"])},
+                )
+            if not self.lti_1p3_tool_public_key and not self.lti_1p3_tool_keyset_url:
+                # Raise ValidationError if public key and keyset URL are missing.
+                validation_errors.update(
+                    {
+                        "lti_1p3_tool_public_key": _(
+                            MESSAGES["required_pubkey_or_keyset"]
+                        ),
+                        "lti_1p3_tool_keyset_url": _(
+                            MESSAGES["required_pubkey_or_keyset"]
+                        ),
+                    },
+                )
+
+        if validation_errors:
+            raise ValidationError(validation_errors)
+
+    def save(self, *args, **kwargs):
+        if self.version == LTIVersion.LTI_1P3:
+            # Generate client ID or private key ID if missing.
+            if not self.lti_1p3_client_id:
+                self.lti_1p3_client_id = str(uuid.uuid4())
+            if not self.lti_1p3_private_key_id:
+                self.lti_1p3_private_key_id = str(uuid.uuid4())
+
+            # Regenerate public JWK.
+            public_keys = jwk.KEYS()
+            public_keys.append(RSAKey(
+                kid=self.lti_1p3_private_key_id,
+                key=RSA.import_key(self.lti_1p3_private_key),
+            ))
+            self.lti_1p3_public_jwk = json.loads(public_keys.dump_jwks())
```

Comment:
Why do we want to regenerate JWK on every save? The `LtiConfiguration` model is doing this only when the value is missing.

Reply:
This is only executed when the external configuration is saved, I don't think there is much of a performance hit compared to verifying if any key is missing every time they are being accessed. The method used on the LtiConfiguration has a disadvantage, if the LtiConfiguration private key is updated, the public JWK will need to be deleted so it's regenerated. I could add more logic to make sure the public JWK is only re-generated if the private key or private key ID is changed, but I didn't find a simple way to do this on the model save method.